obsplus/events/validate.py

```diff
@@ -100,57 +99,191 @@ def check_origins(event: Event):
+def check_p_lims(event: Event, p_lim=None, **kwargs):
+    """
+    Check for P picks that aren't within p_lim of the median pick (if provided)
+    """
+    if p_lim is not None:
+        df = obsplus.picks_to_df(event)
+        df = df.loc[(df.evaluation_status != "rejected") & (df.phase_hint == "P")]
+        med = df.time.median()
+        bad = df.loc[abs(df.time - med) > p_lim]
+        assert len(bad) == 0, (
+            "Outlying P pick found:\n"
+            f"event_id: {event.resource_id.id}, "
+            f"seed_id/s: {bad.seed_id.tolist()}"
+        )
+
+
+@validator("obsplus", Event)
+def check_amp_lims(event: Event, amp_lim=None, **kwargs):
+    """
+    Check for amplitudes that aren't below amp_lim (if provided)
+    """
+    if amp_lim is not None:
+        bad = []
+        for amp in event.amplitudes:
+            if amp.generic_amplitude > amp_lim:
+                wid = amp.waveform_id
+                nslc = (
+                    f"{wid.network_code}.{wid.station_code}."
+                    f"{wid.location_code}.{wid.channel_code}"
+                )
+                bad.append(nslc)
+        assert len(bad) == 0, (
+            "Above limit amplitude found:\n"
+            f"event_id: {event.resource_id.id}, "
+            f"seed_id/s: {bad}"
+        )
+
+
+@validator("obsplus", Event)
+def check_amp_filts(event: Event, filt_amps=None, **kwargs):
+    """
+    Check that all amplitudes have a specified filter id (if provided)
+    """
+    if filt_amps is not None:
+        if type(filt_amps) is ResourceIdentifier:
+            filt_amps = filt_amps.id
+        bad = []
+        bad_filters = []
+        for amp in event.amplitudes:
+            if amp.filter_id.id != filt_amps:
+                wid = amp.waveform_id
+                nslc = (
+                    f"{wid.network_code}.{wid.station_code}."
+                    f"{wid.location_code}.{wid.channel_code}"
+                )
+                bad.append(nslc)
+                if amp.filter_id.id not in bad_filters:
+                    bad_filters.append(amp.filter_id.id)
+        assert len(bad) == 0, (
+            "Unexpected amplitude filter found:\n"
+            f"event_id: {event.resource_id.id}, "
+            f"seed_id/s: {bad}, "
+            f"filters_used: {set(bad_filters)}"
+        )
+
+
+@validator("obsplus", Event)
+def check_z_amps(event: Event, no_z_amps=False, **kwargs):
+    """
+    Check for IAML picks on Z channels (if no_z_amps is True)
+    """
+    if no_z_amps:
+        df = obsplus.picks_to_df(event)
+        df = df.loc[(df.evaluation_status != "rejected") & (df.phase_hint == "IAML")]
+        bad = df.loc[df.channel.str.endswith("Z")].seed_id.tolist()
+        assert len(bad) == 0, (
+            "Amplitude pick on Z axis found:\n"
+            f"event_id: {event.resource_id.id}, "
+            f"seed_id/s: {bad}"
+        )
+
+
+@validator("obsplus", Event)
+def check_amp_times(event: Event, **kwargs):
```

Comment:
From #64: Moment amplitudes, at least as I have implemented them in another code I am working on, are based off of phase picks, but it is possible for the start of the time window used in the moment calculation be different (slightly earlier) than the reported pick time, so the reference time will not match.

Reply:
Sure, fair enough. Should we assert that the pick time is contained in the time window then?